File: 91.decode-ways.py
class Solution:
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s or s[0] == '0':
            return 0
        last = 1
        before_last = 1
        for i in range(1, len(s)):
            cur = 0
            if s[i - 1] == '1' or (s[i - 1] == '2' and s[i] <= '6' and s[i] >= '0'):
                if s[i] != '0':
                    cur = last + before_last
                else:
                    cur = before_last
            elif s[i] == '0':
                return 0
            else:
                cur = last
            before_last = last
            last = cur
        return last

File: 91.decode-ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        if (s.empty() || s[0] == '0') {
            return 0;
        }
        int last = 1;
        int before_last = 1;
        for (int i = 1; i < s.length(); i++) {
            int cur;
            if (s[i - 1] == '1' || (s[i - 1] == '2' && s[i] <= '6' && s[i] >= '0')) {
                cur = s[i] == '0' ? before_last : last + before_last;
            } else if (s[i] == '0') {
                return 0;
            } else {
                cur = last;
            }
            before_last = last;
            last = cur;
        }
        return last;
    }
};